operator/cluster/cluster.go

```diff
@@ -349,39 +351,201 @@ func ScaleDownBase(clientset *kubernetes.Clientset, client *rest.RESTClient, rep
+		}
+
+		// regardless, ensure the limit is gone
+		delete(deployment.Spec.Template.Spec.Containers[0].Resources.Limits, v1.ResourceCPU)
+
+		// handle the memory update
+		if removeMemory {
+			delete(deployment.Spec.Template.Spec.Containers[0].Resources.Requests, v1.ResourceMemory)
+		} else {
+			deployment.Spec.Template.Spec.Containers[0].Resources.Requests[v1.ResourceMemory] = memory
+		}
+
+		// regardless, ensure the limit is gone
+		delete(deployment.Spec.Template.Spec.Containers[0].Resources.Limits, v1.ResourceMemory)
+
+		// Before applying the update, we want to explicitly stop PostgreSQL on each
+		// instance. This prevents PostgreSQL from having to boot up in crash
+		// recovery mode.
+		//
+		// If an error is returned, we only issue a warning
+		if err := stopPostgreSQLInstance(clientset, restConfig, deployment); err != nil {
+			log.Warn(err)
+		}
+
+		// update the deployment with the new values
+		if err := kubeapi.UpdateDeployment(clientset, &deployment); err != nil {
 			return err
 		}
 	}
+
 	return nil
 }
 
-func cleanupPreviousTasks(client *rest.RESTClient, clusterName, namespace string) error {
-
-	selector := config.LABEL_PG_CLUSTER + "=" + clusterName
-	taskList := crv1.PgtaskList{}
+// UpdateTablespaces updates the PostgreSQL instance Deployments to update
+// what tablespaces are mounted.
+// Though any new tablespaces are present in the CRD, to attempt to do less work
+// this function takes a map of the new tablespaces that are being added, so we
+// only have to check and create the PVCs that are being mounted at this time
+//
+// To do this, iterate through the the tablespace mount map that is present in
+// the new cluster.
+func UpdateTablespaces(clientset *kubernetes.Clientset, restConfig *rest.Config,
+	cluster *crv1.Pgcluster, newTablespaces map[string]crv1.PgStorageSpec) error {
+	// first, get a list of all of the instance deployments for the cluster
+	deployments, err := operator.GetInstanceDeployments(clientset, cluster)
 
-	err := kubeapi.GetpgtasksBySelector(client, &taskList, selector, namespace)
 	if err != nil {
 		return err
 	}
 
-	for _, t := range taskList.Items {
-		err = kubeapi.Deletepgtask(client, t.Name, namespace)
+	// now we can start creating the new tablespaces! First, create the new
+	// PVCs. The PVCs are created for each **instance** in the cluster, as every
+	// instance needs to have a distinct PVC for each tablespace
+	for tablespaceName, storageSpec := range newTablespaces {
+		for _, deployment := range deployments.Items {
+			// get the name of the tablespace PVC for that instance
+			tablespacePVCName := operator.GetTablespacePVCName(deployment.Name, tablespaceName)
+
+			log.Debugf("creating tablespace PVC [%s] for [%s]", tablespacePVCName, deployment.Name)
+
+			// and now create it! If it errors, we just need to return, which
+			// potentially leaves things in an inconsistent state, but at this point
+			// only PVC objects have been created
```

Comment:
This was code that I moved from a different commit. Overall, we'd have to make the Operator code more transactional.

This could be a starting point. I think at this point, it'd be safe to delete the PVC or any PVCs that came before it.

Reply:
I think unfortunately I need to pass on this for now, but certainly something to circle back to.